nomad/csi_batch.go

```diff
@@ -0,0 +1,73 @@
+package nomad
+
+import (
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// csiBatchRelease is a helper for any time we need to release a bunch
+// of volume claims at once. It de-duplicates the volumes and batches
+// the raft messages into manageable chunks. Intended for use by RPCs
+// that have already been forwarded to the leader.
+type csiBatchRelease struct {
+	srv    *Server
+	logger log.Logger
+
+	maxBatchSize int
+	seen         map[string]struct{}
+	batches      []*structs.CSIVolumeClaimBatchRequest
+}
+
+func newCSIBatchRelease(srv *Server, logger log.Logger, max int) *csiBatchRelease {
+	return &csiBatchRelease{
+		srv:          srv,
+		logger:       logger,
+		maxBatchSize: max,
+		seen:         map[string]struct{}{},
+		batches:      []*structs.CSIVolumeClaimBatchRequest{{}},
+	}
+}
+
+// add the volume ID + namespace to the deduplicated batches
+func (c *csiBatchRelease) add(vol, namespace string) {
+	id := vol + namespace
```

Comment:
can this suffer collisions? for example `"ab" + "c"` and `"a" + "bc"`?

other places (memdb, i think?) handle this with `vol + "\x00" + namespace` (or some other delimiter that isn't valid for volume ID or namespace name)

Reply:
actually, it seems like this isn't a problem, because `namespace` is always the same in a given `csiBatchRelease` (either, all volumes claims for a single alloc or a single job). in which case, this could be a property on the `csiBatchRelease` struct instead of an argument to `add` and part of the `seen` identifier.

Reply:
I'm pretty sure the `node_endpoint.go` path can have allocations from multiple jobs. It's not a common case because the update from the client happens frequently but if allocations from two different jobs on the same client terminated, I think we'd get one update for them. So we'd need the namespace for each one.

Good call on the collision. Will fix.